src/pointer/ptr.rs

```diff
@@ -0,0 +1,1007 @@
+        /// 8. `ptr` conforms to the validity invariant of
+        ///   [`VALIDITY_INVARIANT`](invariant::Validity).
+        pub(super) const unsafe fn new(
+            ptr: NonNull<T>,
+        ) -> Ptr<'a, T, ALIASING_INVARIANT, ALIGNMENT_INVARIANT, VALIDITY_INVARIANT> {
+            // SAFETY: The caller has promised to satisfy all safety invariants
+            // of `Ptr`.
+            Self { ptr, _lifetime: PhantomData }
+        }
+
+        /// Constructs a `Ptr` from another `Ptr`, possibly with different
+        /// parameterized safety invariants.
+        ///
+        /// # Safety
+        ///
+        /// The caller promises that:
+        /// - `ptr` conforms to the aliasing invariant of
+        ///   [`ALIASING_INVARIANT`](invariant::Aliasing).
+        /// - `ptr` conforms to the alignment invariant of
+        ///   [`ALIGNMENT_INVARIANT`](invariant::Alignment).
+        /// - `ptr` conforms to the validity invariant of
+        ///   [`VALIDITY_INVARIANT`](invariant::Validity).
+        pub(super) const unsafe fn from_ptr<
+            const INITIAL_ALIASING_INVARIANT: invariant::Aliasing,
+            const INITIAL_ALIGNMENT_INVARIANT: invariant::Alignment,
+            const INITIAL_VALIDITY_INVARIANT: invariant::Validity,
+        >(
+            ptr: Ptr<
+                'a,
+                T,
+                INITIAL_ALIASING_INVARIANT,
+                INITIAL_ALIGNMENT_INVARIANT,
+                INITIAL_VALIDITY_INVARIANT,
+            >,
+        ) -> Self {
+            // SAFETY: The caller has promised to satisfy all parameterized
+            // invariants of `Ptr`. `Ptr`'s other invariants are satisfied
+            // by-contract by the source `Ptr`.
+            unsafe { Self::new(ptr.as_non_null()) }
+        }
+
+        /// Converts this `Ptr<T>` to a [`NonNull<T>`].
+        ///
+        /// Note that this method does not consume `self`. Callers should be
+        /// aware that safe code may assume that the aliasing invariants of
+        /// `self` are not violated, and that the `NonNull` may be used in ways
+        /// that violate those assumptions.
+        pub(crate) const fn as_non_null(&self) -> NonNull<T> {
+            self.ptr
+        }
+    }
+}
+
+pub use def::Ptr;
+
+/// The parameterized invariants of a [`Ptr`].
+//
+// These items are named according to the conventions for enums, should be
+// treated as if they are enums (i.e., only referenced by name), and should be
+// re-defined as enums as soon as it is possible to do so.
+#[allow(non_upper_case_globals)]
+#[doc(hidden)]
+pub mod invariant {
+    /// The aliasing invariant of a [`Ptr`][super::Ptr].
+    ///
+    /// Either:
+    /// - [`AnyAliasing`]
+    /// - [`Shared`]
+    /// - [`Unique`]
+    pub type Aliasing = u8;
+
+    /// No aliasing invariant.
+    pub const AnyAliasing: Aliasing = 1 << 0;
+
+    /// The pointer adheres to the aliasing rules of a `&T`.
+    pub const Shared: Aliasing = 1 << 2;
```

Comment:
@joshlf Previously, you commented that you'd like to see the aliasing rules of `&T` inlined here. Are those rules crisply documented anywhere? I imagine we could write something to the effect of:

> The referent of a shared-aliased `Ptr` may be referenced by any number of shared aliased `Ptr`s or references. Safe code may assume that the referent will not be structurally muted, except through interior `UnsafeCell`s (if any).

How does that sound to you?

Reply:
Yeah that sounds roughly good, except:
- Let's also call out that it may *not* be aliased by any mutable references? Even if those references are held by `unsafe` code and never actually used to perform mutation, the validity invariants are up in the air and it might be considered UB at some point in the future. We should probably also ban aliasing by mutable `Ptr`s for symmetry/abundance of caution regarding future memory model changes.
- s/muted/mutated
- What does "structurally" mean?